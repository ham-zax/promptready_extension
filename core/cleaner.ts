// Core cleaner module - orchestrates the cleaning pipeline
// Based on Architecture Section 7 (Core Modules)

import { BoilerplateFilter } from './filters/boilerplate-filters.js';

export interface CleanerOptions {
  mode: 'general' | 'code_docs';
  preserveCodeBlocks: boolean;
  preserveTables: boolean;
  removeHiddenElements: boolean;
}

export interface CleanerResult {
  cleanedHtml: string;
  metadata: {
    originalLength: number;
    cleanedLength: number;
    elementsRemoved: number;
  };
}

export class ContentCleaner {
  
  /**
   * Main cleaning pipeline
   * 1. Apply boilerplate filters
   * 2. Process with Readability (if available)
   * 3. Sanitize with DOMPurify-like logic
   */
  static async clean(html: string, url: string, options: CleanerOptions): Promise<CleanerResult> {
    try {
      const startTime = performance.now();
      const originalLength = html.length;
      
      // Parse HTML into DOM
      const parser = new DOMParser();
      const doc = parser.parseFromString(html, 'text/html');
      
      if (!doc.body) {
        throw new Error('Invalid HTML content');
      }
      
      // Track elements for metrics
      const originalElementCount = doc.body.querySelectorAll('*').length;
      
      // Step 1: Apply boilerplate filters
      this.applyBoilerplateFilters(doc.body, options);
      
      // Step 2: Remove hidden elements if requested
      if (options.removeHiddenElements) {
        BoilerplateFilter.removeHiddenElements(doc.body);
      }
      
      // Step 3: Apply Readability-style cleaning
      const cleanedContent = this.applyReadabilityLikeCleaning(doc.body, options);
      
      // Step 4: Sanitize content
      const sanitizedContent = this.sanitizeContent(cleanedContent, options);
      
      // Step 5: Final cleanup
      BoilerplateFilter.cleanupEmptyElements(sanitizedContent);
      
      const finalElementCount = sanitizedContent.querySelectorAll('*').length;
      const cleanedHtml = sanitizedContent.innerHTML;
      
      const processingTime = performance.now() - startTime;
      console.log(`Content cleaning completed in ${processingTime.toFixed(2)}ms`);
      
      return {
        cleanedHtml,
        metadata: {
          originalLength,
          cleanedLength: cleanedHtml.length,
          elementsRemoved: originalElementCount - finalElementCount,
        },
      };
      
    } catch (error) {
      console.error('Content cleaning failed:', error);
      throw error;
    }
  }
  
  /**
   * Apply boilerplate filtering rules
   */
  private static applyBoilerplateFilters(element: HTMLElement, options: CleanerOptions): void {
    BoilerplateFilter.applyRules(element, options.mode);
  }
  
  /**
   * Apply Readability-style content extraction
   * This is a simplified version focusing on the main content
   */
  private static applyReadabilityLikeCleaning(element: HTMLElement, options: CleanerOptions): HTMLElement {
    const container = element.cloneNode(true) as HTMLElement;
    
    // Remove script and style elements
    const scriptsAndStyles = Array.from(container.querySelectorAll('script, style, noscript'));
    scriptsAndStyles.forEach(el => el.remove());
    
    // Remove elements with very low content density (likely ads or UI)
    this.removeLowContentDensityElements(container, options);
    
    // Promote content-rich elements
    this.promoteContentElements(container, options);
    
    return container;
  }
  
  /**
   * Remove elements with low content density
   */
  private static removeLowContentDensityElements(element: HTMLElement, options: CleanerOptions): void {
    const allElements = Array.from(element.querySelectorAll('div, section, article, aside'));
    
    for (const el of allElements) {
      // Skip if it contains preserved elements
      if (options.preserveCodeBlocks && el.querySelector('pre, code')) {
        continue;
      }
      if (options.preserveTables && el.querySelector('table')) {
        continue;
      }
      
      const textContent = el.textContent || '';
      const linkContent = Array.from(el.querySelectorAll('a')).reduce((acc, link) => acc + (link.textContent || ''), '');
      const textLength = textContent.length;
      const linkLength = linkContent.length;
      
      // Remove elements where links dominate the content (likely navigation)
      if (textLength > 0 && linkLength / textLength > 0.8) {
        el.remove();
        continue;
      }
      
      // Remove elements with very little text content
      if (textLength < 25 && !el.querySelector('img, video, audio, iframe')) {
        el.remove();
      }
    }
  }
  
  /**
   * Promote elements that likely contain main content
   */
  private static promoteContentElements(element: HTMLElement, options: CleanerOptions): void {
    // Look for article, main, or content containers
    const contentContainers = Array.from(element.querySelectorAll('article, main, [role="main"], .content, .post, .entry'));
    
    if (contentContainers.length === 1) {
      // If there's a single clear content container, promote its content
      const container = contentContainers[0];
      const parent = container.parentNode;
      if (parent && parent !== element) {
        // Move content up in the hierarchy
        while (container.firstChild) {
          parent.insertBefore(container.firstChild, container);
        }
        container.remove();
      }
    }
  }
  
  /**
   * Sanitize content for security and cleanliness
   */
  private static sanitizeContent(element: HTMLElement, options: CleanerOptions): HTMLElement {
    const container = element.cloneNode(true) as HTMLElement;
    
    // Remove potentially dangerous elements and attributes
    this.removeDangerousElements(container);
    this.cleanAttributes(container, options);
    
    return container;
  }
  
  /**
   * Remove elements that could be security risks or cause issues
   */
  private static removeDangerousElements(element: HTMLElement): void {
    const dangerousElements = Array.from(element.querySelectorAll(
      'script, object, embed, applet, meta, link[rel="stylesheet"], style'
    ));
    
    dangerousElements.forEach(el => el.remove());
  }
  
  /**
   * Clean and preserve only useful attributes
   */
  private static cleanAttributes(element: HTMLElement, options: CleanerOptions): void {
    // Define allowed attributes per element type
    const allowedAttributes: Record<string, string[]> = {
      'a': ['href', 'title'],
      'img': ['src', 'alt', 'title', 'width', 'height'],
      'video': ['src', 'controls', 'width', 'height'],
      'audio': ['src', 'controls'],
      'table': ['border', 'cellpadding', 'cellspacing'],
      'th': ['colspan', 'rowspan'],
      'td': ['colspan', 'rowspan'],
      'pre': ['class'], // Preserve class for code highlighting
      'code': ['class'], // Preserve class for language identification
    };
    
    // For Code & Docs mode, preserve more attributes on code elements
    if (options.mode === 'code_docs') {
      allowedAttributes['pre'].push('data-language', 'data-filename');
      allowedAttributes['code'].push('data-language');
    }
    
    const allElements = Array.from(element.querySelectorAll('*'));
    
    for (const el of allElements) {
      const tagName = el.tagName.toLowerCase();
      const allowed = allowedAttributes[tagName] || [];
      
      // Remove all attributes except allowed ones
      const attributesToRemove: string[] = [];
      for (let i = 0; i < el.attributes.length; i++) {
        const attr = el.attributes[i];
        if (!allowed.includes(attr.name)) {
          attributesToRemove.push(attr.name);
        }
      }
      
      attributesToRemove.forEach(attrName => {
        el.removeAttribute(attrName);
      });
    }
  }
  
  /**
   * Quick validation of cleaned content
   */
  static validateCleanedContent(html: string): boolean {
    try {
      // Check if HTML is parseable
      const parser = new DOMParser();
      const doc = parser.parseFromString(html, 'text/html');
      
      // Check if there's meaningful content
      const textContent = doc.body?.textContent?.trim() || '';
      if (textContent.length < 10) {
        return false;
      }
      
      // Check for common issues
      const hasScript = doc.querySelector('script') !== null;
      const hasStyle = doc.querySelector('style') !== null;
      
      return !hasScript && !hasStyle;
      
    } catch (error) {
      console.error('Content validation failed:', error);
      return false;
    }
  }
}
