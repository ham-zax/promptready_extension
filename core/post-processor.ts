// Post-processing module for final markdown cleanup and optimization
// Handles structure improvement, formatting consistency, and quality enhancement

export interface PostProcessingOptions {
  cleanupWhitespace: boolean;
  normalizeHeadings: boolean;
  fixListFormatting: boolean;
  removeEmptyLines: boolean;
  maxConsecutiveNewlines: number;
  improveCodeBlocks: boolean;
  enhanceLinks: boolean;
  optimizeImages: boolean;
  addTableOfContents: boolean;
  preserveLineBreaks: boolean;
}

export interface ProcessingResult {
  markdown: string;
  improvements: string[];
  warnings: string[];
  stats: {
    originalLength: number;
    processedLength: number;
    linesRemoved: number;
    structureChanges: number;
  };
}

export class MarkdownPostProcessor {
  
  private static readonly DEFAULT_OPTIONS: PostProcessingOptions = {
    cleanupWhitespace: true,
    normalizeHeadings: true,
    fixListFormatting: true,
    removeEmptyLines: true,
    maxConsecutiveNewlines: 2,
    improveCodeBlocks: true,
    enhanceLinks: true,
    optimizeImages: true,
    addTableOfContents: false,
    preserveLineBreaks: false,
  };

  /**
   * Main post-processing pipeline
   */
  static process(
    markdown: string,
    options: Partial<PostProcessingOptions> = {}
  ): ProcessingResult {
    const config = { ...this.DEFAULT_OPTIONS, ...options };
    const improvements: string[] = [];
    const warnings: string[] = [];
    const originalLength = markdown.length;
    let processed = markdown;
    let structureChanges = 0;

    console.log('[PostProcessor] Starting post-processing...');

    try {
      // Step 1: Basic cleanup
      if (config.cleanupWhitespace) {
        const result = this.cleanupWhitespace(processed, config);
        processed = result.markdown;
        improvements.push(...result.improvements);
      }

      // Step 2: Normalize headings
      if (config.normalizeHeadings) {
        const result = this.normalizeHeadings(processed);
        processed = result.markdown;
        improvements.push(...result.improvements);
        structureChanges += result.changes;
      }

      // Step 3: Fix list formatting
      if (config.fixListFormatting) {
        const result = this.fixListFormatting(processed);
        processed = result.markdown;
        improvements.push(...result.improvements);
      }

      // Step 4: Improve code blocks
      if (config.improveCodeBlocks) {
        const result = this.improveCodeBlocks(processed);
        processed = result.markdown;
        improvements.push(...result.improvements);
      }

      // Step 5: Enhance links
      if (config.enhanceLinks) {
        const result = this.enhanceLinks(processed);
        processed = result.markdown;
        improvements.push(...result.improvements);
        warnings.push(...result.warnings);
      }

      // Step 6: Optimize images
      if (config.optimizeImages) {
        const result = this.optimizeImages(processed);
        processed = result.markdown;
        improvements.push(...result.improvements);
      }

      // Step 7: Remove excessive empty lines
      if (config.removeEmptyLines) {
        const result = this.removeExcessiveEmptyLines(processed, config.maxConsecutiveNewlines);
        processed = result.markdown;
        improvements.push(...result.improvements);
      }

      // Step 8: Add table of contents (if requested)
      if (config.addTableOfContents) {
        const result = this.addTableOfContents(processed);
        processed = result.markdown;
        improvements.push(...result.improvements);
        structureChanges += result.changes;
      }

      // Step 9: Final cleanup
      processed = this.finalCleanup(processed, config);

      const linesRemoved = (markdown.match(/\n/g) || []).length - (processed.match(/\n/g) || []).length;

      console.log(`[PostProcessor] Processing completed. ${improvements.length} improvements made.`);

      return {
        markdown: processed,
        improvements,
        warnings,
        stats: {
          originalLength,
          processedLength: processed.length,
          linesRemoved,
          structureChanges,
        },
      };

    } catch (error) {
      console.error('[PostProcessor] Processing failed:', error);
      warnings.push(`Post-processing failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
      
      return {
        markdown,
        improvements,
        warnings,
        stats: {
          originalLength,
          processedLength: markdown.length,
          linesRemoved: 0,
          structureChanges: 0,
        },
      };
    }
  }

  /**
   * Clean up whitespace issues
   */
  private static cleanupWhitespace(
    markdown: string,
    config: PostProcessingOptions
  ): { markdown: string; improvements: string[] } {
    const improvements: string[] = [];
    let processed = markdown;

    // Remove trailing whitespace
    const beforeTrailing = processed;
    processed = processed.replace(/[ \t]+$/gm, '');
    if (processed !== beforeTrailing) {
      improvements.push('Removed trailing whitespace');
    }

    // Fix mixed line endings
    const beforeLineEndings = processed;
    processed = processed.replace(/\r\n/g, '\n').replace(/\r/g, '\n');
    if (processed !== beforeLineEndings) {
      improvements.push('Normalized line endings');
    }

    // Remove tabs and replace with spaces
    const beforeTabs = processed;
    processed = processed.replace(/\t/g, '  ');
    if (processed !== beforeTabs) {
      improvements.push('Replaced tabs with spaces');
    }

    return { markdown: processed, improvements };
  }

  /**
   * Normalize heading structure and formatting
   */
  private static normalizeHeadings(markdown: string): {
    markdown: string;
    improvements: string[];
    changes: number;
  } {
    const improvements: string[] = [];
    let processed = markdown;
    let changes = 0;

    // Fix heading spacing
    const beforeSpacing = processed;
    processed = processed.replace(/^(#{1,6})\s*(.+)$/gm, '$1 $2');
    if (processed !== beforeSpacing) {
      improvements.push('Fixed heading spacing');
      changes++;
    }

    // Ensure headings have proper line breaks
    const beforeBreaks = processed;
    processed = processed.replace(/^(#{1,6}\s.+)$/gm, '\n$1\n');
    processed = processed.replace(/^\n+/gm, '\n'); // Clean up excessive newlines
    if (processed !== beforeBreaks) {
      improvements.push('Added proper heading line breaks');
      changes++;
    }

    // Fix heading hierarchy (ensure no skipped levels)
    const headingMatches = processed.match(/^(#{1,6})\s/gm);
    if (headingMatches) {
      const levels = headingMatches.map(h => h.length - 1);
      let currentLevel = 0;
      let hierarchyFixed = false;

      for (let i = 0; i < levels.length; i++) {
        const level = levels[i];
        if (level > currentLevel + 1) {
          // Skip detected, fix it
          const newLevel = currentLevel + 1;
          const oldHeading = '#'.repeat(level);
          const newHeading = '#'.repeat(newLevel);
          processed = processed.replace(new RegExp(`^${oldHeading}\\s`, 'm'), `${newHeading} `);
          hierarchyFixed = true;
          currentLevel = newLevel;
        } else {
          currentLevel = level;
        }
      }

      if (hierarchyFixed) {
        improvements.push('Fixed heading hierarchy');
        changes++;
      }
    }

    return { markdown: processed, improvements, changes };
  }

  /**
   * Fix list formatting issues
   */
  private static fixListFormatting(markdown: string): {
    markdown: string;
    improvements: string[];
  } {
    const improvements: string[] = [];
    let processed = markdown;

    // Normalize bullet list markers
    const beforeBullets = processed;
    processed = processed.replace(/^(\s*)[*+]\s+/gm, '$1- ');
    if (processed !== beforeBullets) {
      improvements.push('Normalized bullet list markers');
    }

    // Fix list item spacing
    const beforeSpacing = processed;
    processed = processed.replace(/^(\s*[-*+])\s+/gm, '$1 ');
    if (processed !== beforeSpacing) {
      improvements.push('Fixed list item spacing');
    }

    // Fix numbered list formatting
    const beforeNumbered = processed;
    processed = processed.replace(/^(\s*)(\d+)\.\s+/gm, '$1$2. ');
    if (processed !== beforeNumbered) {
      improvements.push('Fixed numbered list formatting');
    }

    // Ensure proper list line breaks
    const beforeListBreaks = processed;
    processed = processed.replace(/^(\s*[-*+]\s.+)$/gm, '\n$1');
    processed = processed.replace(/^(\s*\d+\.\s.+)$/gm, '\n$1');
    processed = processed.replace(/\n{3,}/g, '\n\n'); // Clean up excessive newlines
    if (processed !== beforeListBreaks) {
      improvements.push('Added proper list line breaks');
    }

    return { markdown: processed, improvements };
  }

  /**
   * Improve code block formatting
   */
  private static improveCodeBlocks(markdown: string): {
    markdown: string;
    improvements: string[];
  } {
    const improvements: string[] = [];
    let processed = markdown;

    // Fix code block fencing
    const beforeFencing = processed;
    processed = processed.replace(/```(\w*)\n([\s\S]*?)\n```/g, (match, lang, code) => {
      const cleanCode = code.replace(/^\n+|\n+$/g, ''); // Remove leading/trailing newlines
      return `\n\n\`\`\`${lang}\n${cleanCode}\n\`\`\`\n\n`;
    });
    if (processed !== beforeFencing) {
      improvements.push('Improved code block formatting');
    }

    // Fix inline code spacing
    const beforeInline = processed;
    processed = processed.replace(/`([^`]+)`/g, (match, code) => {
      const cleanCode = code.trim();
      return `\`${cleanCode}\``;
    });
    if (processed !== beforeInline) {
      improvements.push('Fixed inline code spacing');
    }

    // Detect and fix indented code blocks
    const beforeIndented = processed;
    processed = processed.replace(/^(    .+)$/gm, (match) => {
      // Convert indented code to fenced code blocks
      const code = match.replace(/^    /gm, '');
      return `\n\`\`\`\n${code}\n\`\`\`\n`;
    });
    if (processed !== beforeIndented) {
      improvements.push('Converted indented code to fenced blocks');
    }

    return { markdown: processed, improvements };
  }

  /**
   * Enhance link formatting and validation
   */
  private static enhanceLinks(markdown: string): {
    markdown: string;
    improvements: string[];
    warnings: string[];
  } {
    const improvements: string[] = [];
    const warnings: string[] = [];
    let processed = markdown;

    // Fix link formatting
    const beforeLinks = processed;
    processed = processed.replace(/\[([^\]]*)\]\(([^)]*)\)/g, (match, text, url) => {
      const cleanText = text.trim();
      const cleanUrl = url.trim();
      
      // Validate URL
      if (!cleanUrl) {
        warnings.push(`Empty URL found for link: ${cleanText}`);
        return cleanText; // Remove broken link
      }

      // Fix relative URLs (basic validation)
      if (cleanUrl.startsWith('//')) {
        return `[${cleanText}](https:${cleanUrl})`;
      }

      return `[${cleanText}](${cleanUrl})`;
    });
    if (processed !== beforeLinks) {
      improvements.push('Enhanced link formatting');
    }

    // Fix image links
    const beforeImages = processed;
    processed = processed.replace(/!\[([^\]]*)\]\(([^)]*)\)/g, (match, alt, src) => {
      const cleanAlt = alt.trim();
      const cleanSrc = src.trim();
      
      if (!cleanSrc) {
        warnings.push(`Empty image source found: ${cleanAlt}`);
        return ''; // Remove broken image
      }

      return `![${cleanAlt}](${cleanSrc})`;
    });
    if (processed !== beforeImages) {
      improvements.push('Enhanced image link formatting');
    }

    return { markdown: processed, improvements, warnings };
  }

  /**
   * Optimize image formatting
   */
  private static optimizeImages(markdown: string): {
    markdown: string;
    improvements: string[];
  } {
    const improvements: string[] = [];
    let processed = markdown;

    // Ensure images have proper line breaks
    const beforeBreaks = processed;
    processed = processed.replace(/!\[([^\]]*)\]\(([^)]*)\)/g, '\n\n$&\n\n');
    processed = processed.replace(/\n{3,}/g, '\n\n'); // Clean up excessive newlines
    if (processed !== beforeBreaks) {
      improvements.push('Added proper image line breaks');
    }

    // Add alt text for images without it
    const beforeAlt = processed;
    processed = processed.replace(/!\[\]\(([^)]+)\)/g, (match, src) => {
      const filename = src.split('/').pop()?.split('.')[0] || 'image';
      return `![${filename}](${src})`;
    });
    if (processed !== beforeAlt) {
      improvements.push('Added missing alt text for images');
    }

    return { markdown: processed, improvements };
  }

  /**
   * Remove excessive empty lines
   */
  private static removeExcessiveEmptyLines(
    markdown: string,
    maxConsecutive: number
  ): { markdown: string; improvements: string[] } {
    const improvements: string[] = [];
    
    const beforeCleanup = markdown;
    const pattern = new RegExp(`\\n{${maxConsecutive + 1},}`, 'g');
    const processed = markdown.replace(pattern, '\n'.repeat(maxConsecutive));
    
    if (processed !== beforeCleanup) {
      improvements.push(`Limited consecutive empty lines to ${maxConsecutive}`);
    }

    return { markdown: processed, improvements };
  }

  /**
   * Add table of contents
   */
  private static addTableOfContents(markdown: string): {
    markdown: string;
    improvements: string[];
    changes: number;
  } {
    const improvements: string[] = [];
    let changes = 0;

    // Extract headings
    const headings = markdown.match(/^(#{1,6})\s(.+)$/gm);
    
    if (!headings || headings.length < 2) {
      return { markdown, improvements, changes };
    }

    // Generate TOC
    let toc = '\n## Table of Contents\n\n';
    headings.forEach(heading => {
      const level = heading.match(/^(#{1,6})/)?.[1].length || 1;
      const text = heading.replace(/^#{1,6}\s/, '');
      const anchor = text.toLowerCase().replace(/[^\w\s-]/g, '').replace(/\s+/g, '-');
      const indent = '  '.repeat(Math.max(0, level - 2));
      toc += `${indent}- [${text}](#${anchor})\n`;
    });

    toc += '\n';

    // Insert TOC after first heading
    const firstHeadingIndex = markdown.search(/^#{1,6}\s/m);
    if (firstHeadingIndex !== -1) {
      const nextLineIndex = markdown.indexOf('\n', firstHeadingIndex);
      if (nextLineIndex !== -1) {
        const processed = markdown.slice(0, nextLineIndex + 1) + toc + markdown.slice(nextLineIndex + 1);
        improvements.push('Added table of contents');
        changes = 1;
        return { markdown: processed, improvements, changes };
      }
    }

    return { markdown, improvements, changes };
  }

  /**
   * Final cleanup pass
   */
  private static finalCleanup(markdown: string, config: PostProcessingOptions): string {
    let processed = markdown;

    // Remove non-printable characters
    processed = processed.replace(/[\u0000-\u0008\u000b\u000c\u000e-\u001f\u007f-\u009f\u00ad\u061c\u200b-\u200f\u2028\u2029\ufeff\ufff9-\ufffc]/g, '');

    // Ensure file ends with single newline
    processed = processed.replace(/\n*$/, '\n');

    // Preserve line breaks if requested
    if (config.preserveLineBreaks) {
      processed = processed.replace(/\n/g, '  \n');
    }

    return processed;
  }
}
