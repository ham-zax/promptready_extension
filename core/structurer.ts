// Core structurer module - converts cleaned HTML to structured blocks
// Based on Architecture Section 7 (Core Modules) and PRD Section 14.1

import { PromptReadyExport, ContentBlock, ExportMetadata } from '../lib/types.js';

export interface StructurerOptions {
  mode: 'general' | 'code_docs';
  preserveCodeLanguages: boolean;
  maxHeadingLevel: number;
  includeTableHeaders: boolean;
}

export class ContentStructurer {
  
  /**
   * Convert cleaned HTML to structured PromptReadyExport format
   */
  static async structure(
    cleanedHtml: string,
    metadata: ExportMetadata,
    options: StructurerOptions
  ): Promise<PromptReadyExport> {
    try {
      const parser = new DOMParser();
      const doc = parser.parseFromString(cleanedHtml, 'text/html');
      
      if (!doc.body) {
        throw new Error('Invalid HTML content for structuring');
      }
      
      const blocks = this.extractBlocks(doc.body, options);
      
      return {
        version: '1.0',
        metadata,
        blocks,
      };
      
    } catch (error) {
      console.error('Content structuring failed:', error);
      throw error;
    }
  }
  
  /**
   * Convert structured blocks to Markdown
   */
  static blocksToMarkdown(blocks: ContentBlock[]): string {
    const markdownParts: string[] = [];
    
    for (const block of blocks) {
      const markdown = this.blockToMarkdown(block);
      if (markdown.trim()) {
        markdownParts.push(markdown);
      }
    }
    
    return markdownParts.join('\n\n');
  }
  
  /**
   * Extract structured blocks from DOM
   */
  private static extractBlocks(element: HTMLElement, options: StructurerOptions): ContentBlock[] {
    const blocks: ContentBlock[] = [];
    const childNodes = Array.from(element.childNodes);
    
    for (const node of childNodes) {
      if (node.nodeType === Node.ELEMENT_NODE) {
        const elementNode = node as HTMLElement;
        const block = this.processElement(elementNode, options);
        if (block) {
          blocks.push(block);
        }
      } else if (node.nodeType === Node.TEXT_NODE) {
        const textContent = node.textContent?.trim();
        if (textContent) {
          blocks.push({
            type: 'paragraph',
            text: textContent,
          });
        }
      }
    }
    
    return blocks;
  }
  
  /**
   * Process a single HTML element into a content block
   */
  private static processElement(element: HTMLElement, options: StructurerOptions): ContentBlock | null {
    const tagName = element.tagName.toLowerCase();
    
    switch (tagName) {
      case 'h1':
      case 'h2':
      case 'h3':
      case 'h4':
      case 'h5':
      case 'h6':
        return this.processHeading(element, options);
        
      case 'p':
        return this.processParagraph(element);
        
      case 'ul':
      case 'ol':
        return this.processList(element);
        
      case 'table':
        return this.processTable(element, options);
        
      case 'pre':
        return this.processCodeBlock(element, options);
        
      case 'code':
        // Inline code - treat as part of paragraph unless it's a block
        if (element.parentElement?.tagName.toLowerCase() !== 'pre') {
          return this.processParagraph(element);
        }
        return null;
        
      case 'blockquote':
        return this.processQuote(element);
        
      case 'div':
      case 'section':
      case 'article':
        // For container elements, process their children
        const childBlocks = this.extractBlocks(element, options);
        // Flatten the blocks (don't create nested structure)
        return null; // Will be handled by recursive extraction
        
      default:
        // For other elements, treat as paragraph if they have meaningful text
        const textContent = element.textContent?.trim();
        if (textContent && textContent.length > 0) {
          return {
            type: 'paragraph',
            text: textContent,
          };
        }
        return null;
    }
  }
  
  /**
   * Process heading elements
   */
  private static processHeading(element: HTMLElement, options: StructurerOptions): ContentBlock | null {
    const tagName = element.tagName.toLowerCase();
    const level = parseInt(tagName.charAt(1), 10);
    
    // Respect maxHeadingLevel option
    const adjustedLevel = Math.min(level, options.maxHeadingLevel);
    
    const text = element.textContent?.trim();
    if (!text) return null;
    
    return {
      type: 'heading',
      level: adjustedLevel,
      text,
    };
  }
  
  /**
   * Process paragraph elements
   */
  private static processParagraph(element: HTMLElement): ContentBlock | null {
    let text = '';
    
    // Handle mixed content (text + inline elements)
    for (const node of Array.from(element.childNodes)) {
      if (node.nodeType === Node.TEXT_NODE) {
        text += node.textContent || '';
      } else if (node.nodeType === Node.ELEMENT_NODE) {
        const el = node as HTMLElement;
        const tagName = el.tagName.toLowerCase();
        
        switch (tagName) {
          case 'strong':
          case 'b':
            text += `**${el.textContent}**`;
            break;
          case 'em':
          case 'i':
            text += `*${el.textContent}*`;
            break;
          case 'code':
            text += `\`${el.textContent}\``;
            break;
          case 'a':
            const href = el.getAttribute('href');
            const linkText = el.textContent;
            if (href && linkText) {
              text += `[${linkText}](${href})`;
            } else {
              text += linkText || '';
            }
            break;
          default:
            text += el.textContent || '';
        }
      }
    }
    
    text = text.trim();
    if (!text) return null;
    
    return {
      type: 'paragraph',
      text,
    };
  }
  
  /**
   * Process list elements (ul/ol)
   */
  private static processList(element: HTMLElement): ContentBlock | null {
    const items: string[] = [];
    const listItems = Array.from(element.querySelectorAll('li'));
    
    for (const li of listItems) {
      const text = li.textContent?.trim();
      if (text) {
        items.push(text);
      }
    }
    
    if (items.length === 0) return null;
    
    return {
      type: 'list',
      items,
    };
  }
  
  /**
   * Process table elements
   */
  private static processTable(element: HTMLElement, options: StructurerOptions): ContentBlock | null {
    const headers: string[] = [];
    const rows: string[][] = [];
    
    // Extract headers
    const headerCells = Array.from(element.querySelectorAll('thead th, tr:first-child th, tr:first-child td'));
    
    if (options.includeTableHeaders && headerCells.length > 0) {
      for (const cell of headerCells) {
        headers.push(cell.textContent?.trim() || '');
      }
    }
    
    // Extract data rows
    const dataRows = Array.from(element.querySelectorAll('tbody tr, tr'));
    
    for (const row of dataRows) {
      // Skip header row if we already processed it
      if (headers.length > 0 && row === dataRows[0] && row.querySelector('th')) {
        continue;
      }
      
      const cells = Array.from(row.querySelectorAll('td, th'));
      const rowData: string[] = [];
      
      for (const cell of cells) {
        rowData.push(cell.textContent?.trim() || '');
      }
      
      if (rowData.some(cell => cell.length > 0)) {
        rows.push(rowData);
      }
    }
    
    if (rows.length === 0) return null;
    
    return {
      type: 'table',
      table: {
        headers: headers.length > 0 ? headers : undefined,
        rows,
      },
    };
  }
  
  /**
   * Process code block elements
   */
  private static processCodeBlock(element: HTMLElement, options: StructurerOptions): ContentBlock | null {
    const codeElement = element.querySelector('code') || element;
    const code = codeElement.textContent || '';
    
    if (!code.trim()) return null;
    
    let language = '';
    
    if (options.preserveCodeLanguages) {
      // Try to detect language from class attributes
      const classAttr = codeElement.className || element.className;
      const languageMatch = classAttr.match(/(?:language-|lang-)([a-zA-Z0-9_+-]+)/);
      if (languageMatch) {
        language = languageMatch[1];
      }
      
      // Try data attributes
      if (!language) {
        language = codeElement.getAttribute('data-language') || 
                   element.getAttribute('data-language') || '';
      }
    }
    
    return {
      type: 'code',
      code: code.trim(),
      language: language || undefined,
    };
  }
  
  /**
   * Process blockquote elements
   */
  private static processQuote(element: HTMLElement): ContentBlock | null {
    const text = element.textContent?.trim();
    if (!text) return null;
    
    return {
      type: 'quote',
      text,
    };
  }
  
  /**
   * Convert a single block to Markdown
   */
  private static blockToMarkdown(block: ContentBlock): string {
    switch (block.type) {
      case 'heading':
        const level = block.level || 1;
        const hashes = '#'.repeat(level);
        return `${hashes} ${block.text}`;
        
      case 'paragraph':
        return block.text || '';
        
      case 'list':
        return (block.items || []).map(item => `- ${item}`).join('\n');
        
      case 'table':
        return this.tableToMarkdown(block.table);
        
      case 'code':
        const language = block.language || '';
        return `\`\`\`${language}\n${block.code}\n\`\`\``;
        
      case 'quote':
        return `> ${block.text}`;
        
      default:
        return '';
    }
  }
  
  /**
   * Convert table structure to Markdown table
   */
  private static tableToMarkdown(table: { headers?: string[]; rows: string[][] } | undefined): string {
    if (!table || !table.rows.length) return '';
    
    const lines: string[] = [];
    
    // Add headers if available
    if (table.headers && table.headers.length > 0) {
      lines.push(`| ${table.headers.join(' | ')} |`);
      lines.push(`| ${table.headers.map(() => '---').join(' | ')} |`);
    }
    
    // Add data rows
    for (const row of table.rows) {
      lines.push(`| ${row.join(' | ')} |`);
    }
    
    return lines.join('\n');
  }
  
  /**
   * Generate citation footer for exports
   */
  static generateCitationFooter(metadata: ExportMetadata): string {
    const capturedDate = new Date(metadata.capturedAt).toLocaleDateString();
    const capturedTime = new Date(metadata.capturedAt).toLocaleTimeString();
    
    return `---

**Source:** ${metadata.url}
**Title:** ${metadata.title}
**Captured:** ${capturedDate} at ${capturedTime}
**Selection Hash:** ${metadata.selectionHash}`;
  }
}
