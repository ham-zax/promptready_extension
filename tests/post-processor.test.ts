import { describe, it, expect } from 'vitest';
import { MarkdownPostProcessor } from '../core/post-processor';

describe('MarkdownPostProcessor', () => {
  it('normalizes headings and limits newlines', () => {
    const input = `###Heading\n\n\n\nText`;
    const { markdown, improvements } = MarkdownPostProcessor.process(input, { normalizeHeadings: true, removeEmptyLines: true, maxConsecutiveNewlines: 2 });
    // Accept H1 normalization at top or retained level with corrected spacing
    expect(markdown).toMatch(/^# Heading|^### Heading/m);
    expect(markdown).not.toMatch(/\n{3,}/);
    expect(improvements.length).toBeGreaterThan(0);
  });

  it('adds TOC when enabled', () => {
    const input = `# Title\n\n## A\nText\n\n### B`;
    const { markdown } = MarkdownPostProcessor.process(input, { addTableOfContents: true });
    expect(markdown).toContain('## Table of Contents');
    expect(markdown).toContain('- [A](#a)');
  });
});

