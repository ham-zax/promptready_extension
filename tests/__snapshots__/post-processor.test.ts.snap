// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MarkdownPostProcessor > adds TOC when enabled 1`] = `
"
# Title

## Table of Contents

- [Title](#title)
- [A](#a)
  - [B](#b)


## A

Text

### B
"
`;
